starknet-e2e-test/ethereum_e2e_settlement.rs

```diff
@@ -0,0 +1,176 @@
+#[tokio::test]
+async fn works_with_initialized_contract(ctx: &mut Context) -> Result<(), anyhow::Error> {
+    // Troubleshooting:
+    // RUST_LOG=mc_settlement=trace MADARA_LOG=1 cargo test --package starknet-e2e-test
+    // works_with_initialized_contract -- --nocapture
+
+    // At this point we have:
+    //   * spawned Ethereum sandbox
+    //   * deployed settlement contract (not initialized yet)
+    //   * temp Madara path with correct ethereum config
+    ctx.initialize(InitData::sn_goerli()).await;
+
+    let mut madara = ctx.launch_madara().await;
+
+    madara.create_n_blocks(3).await?;
+    sleep(Duration::from_millis(300)).await;
+
+    assert_eq!(ctx.read_state().await, StarknetState { block_number: 3u64.into(), state_root: 0u64.into() });
+
+    Ok(())
+}
+
+#[test_context(Context)]
+#[rstest]
+#[tokio::test]
+async fn recovers_from_non_initialized_state(ctx: &mut Context) -> Result<(), anyhow::Error> {
+    let mut madara = ctx.launch_madara().await;
+
+    madara.create_empty_block().await?;
+    // Give the client thread some time to handle the finalized block
+    sleep(Duration::from_millis(100)).await;
+
+    ctx.initialize(InitData::sn_goerli()).await;
+
+    madara.create_empty_block().await?;
+    // Give the client thread some time to recover
+    sleep(Duration::from_millis(100)).await;
+
+    madara.create_empty_block().await?;
+    sleep(Duration::from_millis(100)).await;
+
+    assert_eq!(ctx.read_state().await, StarknetState { block_number: 3u64.into(), state_root: 0u64.into() });
+
+    Ok(())
+}
+
+#[test_context(Context)]
+#[rstest]
+#[tokio::test]
+async fn catches_up_with_the_state_in_the_future(ctx: &mut Context) -> Result<(), anyhow::Error> {
+    ctx.initialize(InitData { block_number: 1u64.into(), ..InitData::sn_goerli() }).await;
+
+    let mut madara = ctx.launch_madara().await;
+
+    // Unless state root calculation is enabled (not by default), we should be fine
+    madara.create_n_blocks(2).await?;
+    // Give the client thread some time to handle the finalized block
+    sleep(Duration::from_millis(200)).await;
+
+    assert_eq!(ctx.read_state().await, StarknetState { block_number: 2u64.into(), state_root: 0u64.into() });
+
+    Ok(())
+}
+
+#[test_context(Context)]
+#[rstest]
+#[tokio::test]
+async fn fails_with_inconsistent_state_in_the_future(ctx: &mut Context) -> Result<(), anyhow::Error> {
+    ctx.initialize(InitData { block_number: 1u64.into(), state_root: 12345u64.into(), ..InitData::sn_goerli() }).await;
+
+    let mut madara = ctx.launch_madara().await;
+
+    madara.create_empty_block().await?;
+    // Give the client thread some time to handle the finalized block
+    sleep(Duration::from_millis(100)).await;
+
+    // Expected connection refused because Madara is shut down at this point
+    let is_connection_refused =
+        assert_matches!(madara.create_empty_block().await, Err(err) => err.downcast_ref::<reqwest::Error>().is_some());
+    assert!(is_connection_refused);
```

Comment:
why are we asserting the result of a result?

Reply:
replaced by `    assert_matches!(madara.create_empty_block().await, Err(err) => assert!(err.downcast_ref::<reqwest::Error>().is_some()));`
we want to be sure that it fails because of a request error, but we have an `anyhow` error, so we need to downcast it